components/constellation/constellation.rs

```diff
@@ -1572,6 +1573,65 @@ where
             EmbedderToConstellationMessage::EmbedderControlResponse(id, response) => {
                 self.handle_embedder_control_response(id, response);
             },
+
+            EmbedderToConstellationMessage::CreateEntangledMessagePorts(
+                router_id,
+                sender,
+                port1,
+                port2,
+            ) => {
+                trace!("creating new entangled ports {port1:?} {port2:?}");
+                self.handle_new_messageport_router(router_id, sender);
+                self.handle_new_messageport(router_id, port1);
+                self.handle_new_messageport(router_id, port2);
+                self.handle_entangle_messageports(port1, port2);
+            },
+
+            EmbedderToConstellationMessage::PostMessageToWebView(
+                webview_id,
+                data,
+            ) => {
+                self.process_any_message_ports(&data);
+                self.handle_post_message_from_constellation(webview_id, data);
+            },
+
+            EmbedderToConstellationMessage::PostMessageToMessagePort(port_id, data) => {
+                self.process_any_message_ports(&data);
+                let task = PortMessageTask {
+                    origin: ImmutableOrigin::new_opaque(),
+                    data: PostMessageData::Serialized(data),
+                };
+                self.handle_reroute_messageport(port_id, task);
+            },
+        }
+    }
+
+    fn process_any_message_ports(&mut self, data: &JSValue) {
+        fn extract_port_ids(value: &JSValue) -> Vec<MessagePortId> {
+            match value {
+                JSValue::Undefined |
+                JSValue::Null |
+                JSValue::Boolean(..) |
+                JSValue::Number(..) |
+                JSValue::String(..) |
+                JSValue::Element(..) |
+                JSValue::ShadowRoot(..) |
+                JSValue::Frame(..) |
+                JSValue::Window(..) => vec![],
+                JSValue::Array(array) => array
+                    .iter()
+                    .flat_map(|value| extract_port_ids(value))
+                    .collect(),
+                JSValue::Object(object) => object
+                    .values()
+                    .flat_map(|value| extract_port_ids(value))
+                    .collect(),
+                JSValue::MessagePort { id, .. } => vec![*id],
+            }
+        }
+
+        for port_id in extract_port_ids(data) {
+            self.handle_messageport_shipped(port_id);
```

Comment:
Seems like you could just send a `shippped` message to deal with this, like the globals currently do.